### DISCUSSION TITLE
Naming Artifacts after committed files

### DISCUSSION THREAD

**Discussion by cyb3rko** [community/community#24947] [Actions]
<p>Hey there.<br>
Yesterday I started using GitHub Actions so I’m very new to this.</p>
<p>My first action works very well:<br>
If I commit a .tex file, it builds it into a PDF file and uploads it as an artifact.</p>
<pre><code class="lang-auto">name: Build LaTeX PDF files
on:
  push:
    paths:
    - '**.tex'

  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Set up Git repository
        uses: actions/checkout@v2
      - name: Read committed files
        uses: jitterbit/get-changed-files@v1
        id: files
        with:
          format: space-delimited
      - name: Print files
        id: print_files
        run: |
          for changed_file in ${{ steps.files.outputs.all }}; do
            echo "Found file: ${changed_file}."
          done
      - name: Compile LaTeX document
        uses: xu-cheng/latex-action@v2
        with:
          root_file: ${{ steps.files.outputs.all }}
      - name: Save PDF file
        uses: actions/upload-artifact@v2
        with:
          name: PDF
          path: test.pdf
</code></pre>
<p>This only works if the .tex file has the name “test.tex” because the upload-artifact action is hardcoded to use the “test.pdf” file.</p>
<p>Now my question is if there is any way to name files dynamically after the name of the .tex file because I read that substrings aren’t a thing in yml and I want to name the pdf after the tex file.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Technically you could reference the output of another step, like you do in the “Print files” step with <code>${{ steps.files.outputs.all }}</code>.</p>
<p>However, as I understand the workflow there could be a list of files, right? If you want all of them it would be easier to tell <code>actions/upload-artifact</code> to grab <em>all</em> PDF files:</p>
<pre><code class="lang-auto">      - name: Save PDF file
        uses: actions/upload-artifact@v2
        with:
          name: PDF
          path: |
            *.pdf
            **/*.pdf
</code></pre>

**Comment by cyb3rko** [comment]
<p>Oh wow, I didn’t even think about this notation: <code>*.pdf</code></p>
<ol>
<li>What does <code>**/*.pdf</code> mean?</li>
<li>The first part of the script is able to work with multiple files but the part where the latex action is called gets a filename as input. Is there any way to extract every string of the list <code>${{ steps.files.outputs.all }}</code> and run following code on every item:</li>
</ol>
<pre><code class="lang-auto">- name: Compile LaTeX document
    uses: xu-cheng/latex-action@v2
    with:
       root_file: filename
</code></pre>
<p>Thanks for your help!</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="cyb3rko" data-post="3" data-topic="163183">
<div class="title">
<div class="quote-controls"></div>
 cyb3rko:</div>
<blockquote>
<p>What does <code>**/*.pdf</code> mean?</p>
</blockquote>
</aside>
<p>The <code>**</code> matches any number of directories, so it would find any PDF files you might have in subdirectories (e.g. <code>a/b/c.pdf</code>). If you don’t have any of those or don’t want them in the artifact you can just remove that part.</p>
<aside class="quote no-group" data-username="cyb3rko" data-post="3" data-topic="163183">
<div class="title">
<div class="quote-controls"></div>
 cyb3rko:</div>
<blockquote>
<p>Is there any way to extract every string of the list <code>${{ steps.files.outputs.all }}</code> and run following code on every item</p>
</blockquote>
</aside>
<p>There isn’t any looping on the step level, but the README for the <a href="https://github.com/xu-cheng/latex-action"><code>xu-cheng/latex-action</code> action</a> says you can give multiple files as a multi-line string (which might require some formatting).</p>
<p>Or you could install the necessary packages on the runner (Ubuntu has <code>texlive</code> packages) and run <code>pdflatex</code> or <code>lualatex</code> in a shell loop.</p>

**Comment by cyb3rko** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="163183">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The <code>**</code> matches any number of directories, so it would find any PDF files you might have in subdirectories (e.g. <code>a/b/c.pdf</code> ). If you don’t have any of those or don’t want them in the artifact you can just remove that part.</p>
</blockquote>
</aside>
<p>Well I think that can be helpful.</p>
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="163183">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>There isn’t any looping on the step level, but the README for the <a href="https://github.com/xu-cheng/latex-action" rel="noopener nofollow ugc"> <code>xu-cheng/latex-action</code> action</a> says you can give multiple files as a multi-line string (which might require some formatting).</p>
</blockquote>
</aside>
<p>Yeah I also found that but how do I format the output list to something like that:</p>
<pre><code class="lang-auto">root_file: |
      file1.tex
      file2.tex
</code></pre>

**Comment by airtower-luna** [comment]
<p>I just did some experimenting, based on hints from this thread:</p><aside class="quote quote-modified" data-post="1" data-topic="16852">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/set-output-truncates-multiline-strings/16852">set-output Truncates Multiline Strings</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    Hi, 
I am using this code to create a release for one of our repositories: 
- shell: bash
        id: release_description
        run: |
          description=$(./resources/get_release_description.sh ${{ steps.versioning.outputs.cli_version }})
          echo $description
          echo "::set-output name=description::$description"
        if: startsWith(steps.commit_message.outputs.commit_message, 'Version change')
      - shell: bash
        run: |
          echo ${{ steps.release_description.…
  </blockquote>
</aside>

<p>The key is that to have newlines in an output you need to URL-encode them. I’ve been using Python to do the space-to-newline conversion because <code>shlex</code> makes it easy to handle escapes (e.g. a file name containing a space) when splitting the space-separated list:</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/airtower-luna/hello-github-actions/blob/ccaec46bba8f11f342b8ba695f795b830d8d7e57/.github/workflows/multiline-output.yaml#L14-L24" target="_blank" rel="noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/hello-github-actions/blob/ccaec46bba8f11f342b8ba695f795b830d8d7e57/.github/workflows/multiline-output.yaml#L14-L24" target="_blank" rel="noopener">airtower-luna/hello-github-actions/blob/ccaec46bba8f11f342b8ba695f795b830d8d7e57/.github/workflows/multiline-output.yaml#L14-L24</a></h4>
<pre class="onebox"><code class="lang-yaml"><ol class="start lines" start="14" style="counter-reset: li-counter 13 ;">
<li>- name: make list newline separated</li>
<li>  id: lines</li>
<li>  run: |</li>
<li>    import os</li>
<li>    import shlex</li>
<li>    # %0A is a URL-encoded newline (\n)</li>
<li>    output = '%0A'.join(shlex.split(os.environ['INPUT_LIST']))</li>
<li>    print(f'::set-output name=list::{output}')</li>
<li>  shell: python3 {0}</li>
<li>  env:</li>
<li>    INPUT_LIST: ${{ steps.files.outputs.list }}</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>You could use something similar to convert the list you have. :slightly_smiling_face:</p>

**Comment by cyb3rko** [comment]
<p>Thanks a lot for investing time in uploading an example, I finally got it working.</p>
<p>Thanks for your time!</p>
